plugins/cache-ehcache/src/main/java/org/opensearch/cache/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cache.keystore;
+
+import org.opensearch.common.cache.keystore.KeyLookupStore;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.core.common.unit.ByteSizeValue;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * This class implements KeyLookupStore using a roaring bitmap with a modulo applied to values.
+ * The modulo increases the density of values, which makes RBMs more memory-efficient. The recommended modulo is ~2^28.
+ * It also maintains a hash set of values which have had collisions. Values which haven't had collisions can be
+ * safely removed from the store. The fraction of collided values should be low,
+ * about 0.5% for a store with 10^7 values and a modulo of 2^28.
+ * The store estimates its memory footprint and will stop adding more values once it reaches its memory cap.
+ */
+public class RBMIntKeyLookupStore implements KeyLookupStore<Integer> {
+    /** Used in settings to distinguish between keystore types. */
+    public static final String KEYSTORE_NAME = "rbm";
+
+    /**
+     * An enum representing modulo values for use in the keystore
+     */
+    public enum KeystoreModuloValue {
+        /** No modulo applied */
+        NONE(0),
+        /** 2^31 */
+        TWO_TO_THIRTY_ONE((int) Math.pow(2, 31)),
+        /** 2^29 */
+        TWO_TO_TWENTY_NINE((int) Math.pow(2, 29)),
+        /** 2^28, recommended value */
+        TWO_TO_TWENTY_EIGHT((int) Math.pow(2, 28)),
+        /** 2^26 */
+        TWO_TO_TWENTY_SIX((int) Math.pow(2, 26));
+
+        private final int value;
+
+        KeystoreModuloValue(int value) {
+            this.value = value;
+        }
+
+        /** Get the numerical value. */
+        public int getValue() {
+            return this.value;
+        }
+    }
+
+    /** The modulo applied to values before adding into the RBM */
+    protected final int modulo;
+    private final int modulo_bitmask;
+    // Since our modulo is always a power of two we can optimize it by ANDing with a particular bitmask
+    KeyStoreStats stats;
+    private RoaringBitmap rbm;
+    private HashMap<Integer, CounterMetric> collidedIntCounters;
+    private HashMap<Integer, HashSet<Integer>> removalSets;
```

Comment:
This might not be friendly for the CPU cache as default Java implementations use Linked List for collisions and then convert into a RedBlack Tree after a certain threshold. 

How about using another global RBM to handle collisions ?